OEP-0055, Project Maintainers (#290)
openedx/open-edx-proposals: oeps/processes/oep-0055-proc-project-maintainers.rst

   :widths: 25 75

   * - OEP
     - :doc:`OEP-0055 <oep-0055-proc-project-maintainers>`
   * - Title
     - Project Maintainers
   * - Last Modified
     - 2022-01-28
   * - Authors
     - Edward Zarecor <e0d@tcril.org>
   * - Arbiter
     - Jeremy Bowman <jbowman@2u.com>
   * - Status
     - Draft
   * - Type
     - Process
   * - Created
     - 2022-01-28
   * - Review Period
     - <start - target end dates for review>
   * - Resolution
     - <links to any discussions where the final status was decided>
   * - References
     - <links to any other relevant discussions or relevant related materials>

Abstract
========

The Open edX platform is a service oriented platform for authoring and
delivering online learning based on the `Django framework`_.  At the
highest level, the platform is composed of numerous services (IDAs
[#]_), micro-front ends (MFEs [#]_) and libraries.  The code for these
services and libraries is managed in git and hosted on GitHub
platform.

Each component part of the Open edX platform requires maintenance to
ensure that it meets, and continues to meet, the project definition of
high quality.  That definition will include project standards,

> appropriate documentation, and fitness for use, among other things.

Reviewer: Should this be better defined? What is appropriate documentation?

- Installation instructions?
- Purpose?
- ADRs?
- Overview diagrams?

Reviewer 2: Expanding on that... if a maintainer team has internal documentation about operations and maintenance is that okay?  Does all documentation related to these need to be made community public?

Reviewer 3: @schenedx I think all of that is appropriate.

@jristau1984 I think enough documentation should be public that people have a sense of how the component is maintained, frequency of upgrades, etc. For a few reasons: knowledge sharing (if you do a good job maintaining, others should learn from you!!); confidence that the component is maintained; transparency. I would err to the side of things should be private only if _absolutely essential_ to 2U's business, and not impacting the community.

Reviewer: @sarina I understand the spirit of helping others and build a strong community. That said, i do believe making code public and knowledge sharing are two ends of open source spectrum. It's easy to put the repo in a public place so others can read and use on their own. It's significant work to make the repo easily understood and knowledge shared and work out of the box. That takes time, process, and effort. The better definition we have on the responsibility of the maintainer, the better 2U can evaluate what are the repositories are we equipped and motivated to make manage them in the open source community. Therefore, we have a trade off in the spirit of this document: Do we want more code (untested, unmaintained and undocumented) in the community? Or do we want more solutions that solve similar problems (With instructions, API docs and knowledge) in the community? It could be somewhere in the middle, and I think we can adjust what that middle is through the responsibility described in the doc.

Reviewer 3: @schenedx understood, and that makes sense. Documentation to me includes things like maintained READMEs - for MFEs, this would include documenting variables used and how to start up a local dev env (I made an [example README for MFEs](https://github.com/openedx/frontend-template-application/blob/master/README-template-frontend-app.rst) that I'm not sure anyone is following - this to me feels like a level of documentation that shouldn't be too hard to audit for correctness periodically, and is helpful to developers).

Reviewer 3: Hi @schenedx @jristau1984 I'd like to resolve this for the time being. Can we enumerate "must-haves" in documentation and "suggested to-haves"? My suggestions are:

### Must-Haves
* A README that explains what the component is used for and as much detail as available such as:
  * what the component does
  * who the maintainers are
  * How to set up a dev env
  * Relevant env variables
  * Document if the code is in an experimental place, is not taking contributions, or anything else relevant to people reading your code.

### Suggested To-Haves
* ADRs
* Architecture Diagrams
* Roadmap
* Issue or project board

Specification
=============

Definitions
-----------

The term **repository** will mean a repository on the GitHub social
coding platform.

The term **component** will typically mean the unit of software housed
in a single GitHub repository.  Exceptions will be made in the case of
edx-platform where there may be multiple distinguishable components in
that single repository.  Components will typically, but not always, be
hosted at the `openedx GitHub organization
<https://github.com/openedx/https://github.com/openedx/>`_.
Components will be "part of" or "critical to" the open-source
platform.

The term **maintainer** will mean a named person or persons
responsible for the health of the component and the component's community.
The remainder of this specification will be focused on the details of
the role.

The term **core contributor** will mean a named person participating
in the `Core Contributor Program`_ as a coding core contributor and
meeting the requirements of that program.

The term **maintainer scrum-of-scrums** will mean the group of all
maintainers who are responsible for flagging issues to project
leadership and proposing process improvements for component
maintenance.

> The term **Technical Oversight Committee** (TOC) will mean the group

Reviewer: This is the only instance where TOC is mentioned. Why? And if no where else in doc, why do we need to define this organization?

Author: The TOC is ultimately the arbiter of last resort for governance matters.  I should make sure that that is perfectly clear here.

Reviewer 4: These are good details to include. Since the TOC isn't referenced anywhere else, though, can you move this out of the definitions section?

Reviewer 3: I think this point is a bit of a nit. It's important to mention that the TOC is the arbiter of last resort for governance matters.

Could we resolve this comment? If not, would you be able to float a specific suggestion?

Reviewer 4: @sarina @e0d Honestly, I'm not sure how the TOC fits into this OEP, so I wouldn't know what to suggest. Can we just remove the definition?

If we're going to mention the TOC in an OEP, it may as well be OEP-1, since their governance-of-last-resort role applies to everything about Open edX, not just maintainership.

Reviewer 3: Personally I do think this is worth mentioning; it's pertinent to the maintainers program to understand the governance structure, and we have a lot of OEPs with a lot of words - I don't think there's a problem with repeating this, especially as it is new -- people who may already have read OEP-1 may not have read a governance update.

What about something like

### Governance Structure

The Maintainers program will be run by the maintainers themselves, working together with input from the broader community. The Program Administrators (link to existing section at end) will serve as a backstop for any issues that arise, and the Technical Oversight Committee (move this section to the end & link to it) will be the arbiter of last resort for governance matters.

Reviewer 3: I decided to leave the definition here, because it would be structurally a bit weird to have definitions at the top and bottom. But now there's that "Governance Structure" section at the end, so TOC is referenced.

Reviewer 4: 💯

  component
* The ability to assess the appropriateness of proposed changes to or
  extensions of the component
* The role of the component in the overall platform and how it
  interacts with other components

It is acceptable for a component to be maintained by a team or group
as long as that arrangement does not reduce the reachablility of the
maintainer or reduce accountability.

Prerequisites
-------------
In order to be a repository maintainer, you must

* Be an active member of the Open edX community who has made technical
  contributions to the project in the last 6 months.  The types of
  contributions that are relevant will match those defined in the
  `code contributors`_
* Abide by the `Open edX Code of Conduct
  <https://openedx.org/code-of-conduct/>`_
* Be a core contributor for the repository in question who has
  demonstrated excellence in the role of core contributor.  From
  time-to-time, in the case of a strong candidate and project need,
  core contributor status and maintainer status may be conferred at
  the same time.

Bootstrapping
-------------
In order to ensure an orderly roll-out of the formal maintainer role,

> existing 2U staff in the owner role will transition into the role of

Reviewer 5: I'm wondering aloud if we should make this a more explicit choice? That is, the current informal owners (both 2U staff and other community members) can:

1. Give up maintainer-ship of the component, so that someone else can pick it up
OR
2. Bring the component up to standard within 1 month

We might also consider giving an extra long grace-period for bringing components up to standard as part of the bootstrapping process in order to give the prior informal owners time to prioritize the needed documentation work.

Author: I think this makes sense generally.  However, if a large number of components were put up for adoption that would not lead to an orderly transition.  I would say:

1. If there are components that are unloved, but owned, I think the current maintainer should see them through the DEPR process.
2. If there are components that are not a high priority for a given team, we should have a process for seeking a new maintainer, before they are "given up."

Fully agree that we should have a defined timeframe for conformance.  However, I also don't expect us to add a ton of new requirements now.

Reviewer 3: I think things like rollout plans and conformance timelines should appear in other documents (probably wiki pages, possibly ADRs in certain repos) rather than codified in an OEP. OEPs aren't meant to change super frequently, and we'd want to keep timelines and execution plans more agile.

Reviewer 6: Do we have a census of the repos/components and how well they meet the requirements here, or a tool to validate a repo?

The repo health spreadsheet seems to check for some of these along with a bunch of others, so it was hard for me to tell how well a given repo meets this standard.

Reviewer 6: Also, are there recommended practices for these standards?  E.g. how to keep the changelog up-to-date and what level of detail belongs there, recommended answers for how to report issues and where to go for help, etc.

Reviewer 3: @e0d I recommend we remove this "Bootstrapping" section into a wiki page. That wiki page can be linked in this OEP, but debating and iterating on the specifics of the pilot and subsequent phases doesn't need to take place in this PR.

contacting a maintainer.

Jobs of the Maintainer
----------------------
Many of the day-to-day specifics of the maintainer role are outside of
the scope of this document.  For example, here we will note that
maintainers are responsible for ensuring that repositories meet
project standards, but not the details of those standards.  Those
standards will be codified in an OEP over time, but will initially be
defined in the wiki page `requirements for public repositories`_ to
allow rapid iteration.

The key aspects of the maintainer role fall into the following
categories

* **Community Stewardship**: Maintainers are representatives of the
  project in the community and are responsible for contributing to an
  environment that is welcoming, vibrant, and growing.
* **Project Management**: Maintainers are responsible for ensuring
  that critical maintenance is done, but this doesn't mean doing it all
  themselves. Effective delegation is an important maintainer skill.
* **Quality Assurance**: Maintainers are responsible for the quality
  of their components.  Quality will mean meeting defined project
  standards.  It can also mean improving our standards or improving
  our processes.

> * **Technical Vision**: Maintainers should have a documented technical

Reviewer: Is this the only documentation we need?

Author: We'll certainly update our definitions over time, so I don't think this is carved in stone.

Do you have particular things in mind that you believe should be captured here?

Reviewer 6: Will there be a standard place and/or format for the "technical vision" document(s)?

Author: I think so, but probably outside the scope of this OEP.  I think it would either need to be part of a standardized README or linked to from and standardized README.

Reviewer 3: @schenedx I think this is tied with my above comment (L50 about appropriate documentation). What would you like to do with this and the above comment?

Reviewer: @sarina I think you should list what you commented from above in this document here as an example. I really like how the documents at minimum should be the readme doc, but encourage other docs to exist as well on the repo. That example should be spelled out here so others can follow, even if they won't be matching the example 100%

Reviewer 3: Thanks! Done!

===========================
OEP-55: Project Maintainers
===========================
.. list-table::
   :widths: 25 75

   * - OEP
     - :doc:`OEP-0055 <oep-0055-proc-project-maintainers>`
   * - Title
     - Project Maintainers
   * - Last Modified
     - 2022-03-16
   * - Authors
     - Edward Zarecor <e0d@tcril.org>, Sarina Canelake <sarina@tcril.org>
   * - Arbiter
     - Jeremy Bowman <jbowman@2u.com>
   * - Status
     - Draft
   * - Type
     - Process
   * - Created
     - 2022-01-28
   * - Review Period
     - 2022-03-21 - 2022-04-11
   * - Resolution
     - <links to any discussions where the final status was decided>
   * - References
     - <links to any other relevant discussions or relevant related materials>

> 

Reviewer 3: ```suggestion

.. contents::
   :local:
   :depth: 3

```

Reviewer 3: Adds a table of contents to the rendered doc

Abstract
========

The Open edX platform is a service oriented platform for authoring and
delivering online learning based on the `Django framework`_.  At the
highest level, the platform is composed of numerous services (IDAs

> [#]_), micro-front ends (MFEs [#]_) and libraries.  The code for these

Reviewer 3: ```suggestion
[#ida]_), micro-front ends (MFEs [#mfe]_) and libraries.  The code for these
```

> [#]_), micro-front ends (MFEs [#]_) and libraries.  The code for these

Author: @sarina  I believe the changes you submitted addressed these?

services and libraries is managed in git and hosted on GitHub

> platform.

Reviewer 6: Nit: "...and hosted on the GitHub platform" or "...and hosted on GitHub's platform"

Author: Fixed.

Each component part of the Open edX platform requires maintenance to
ensure that it meets, and continues to meet, the project definition of
high quality.  That definition will include project standards,
appropriate documentation, and fitness for use, among other things.
That definition will evolve as project needs do and we learn what
works and what doesn't.

Component maintenance is work.  To ensure that that work is done, a
person or persons must be recognized as responsible for doing it,
and held accountable for its performance.

The Open edX project, like many projects, calls this role
"maintainer."  The notion of a project maintainer is typical of

> open-source projects and widely considered an best practice [#]_.

Reviewer 3: ```suggestion
open-source projects and widely considered a best practice [#maintainer]_.
```

> open-source projects and widely considered an best practice [#]_.

Reviewer 6: Nit: "...considered a best practice"

Author: Fixed.

The maintainer role is a leadership role that encompasses both
technical and community management elements.  Both dimensions of the role are
equally important.

> All, non-deprecated, components of the Open edX project must have one

Reviewer 3: ```suggestion
All non-deprecated components of the Open edX project must have one
```

> or more named maintainers.  When no such maintainers exists, this will

Reviewer 3: ```suggestion
or more named maintainers.  When no such maintainer(s) exists, this will
```

be considered an urgent problem to solve.

This OEP proposes a formal definition of the role of maintainer that

> will open to all qualified Open edX community members.  Maintainers

Reviewer 7: ```suggestion
will be open to all qualified Open edX community members.  Maintainers
```

will be responsible for the health of platform components, but not for
their performance in any particular community member's production

> environment[#]_.

Reviewer 3: ```suggestion
environment [#ops]_.
```

> environment[#]_.

Reviewer 6: Nit: I think this footnote marker needs a space; GitHub at least doesn't render it: 
<img width="428" alt="image" src="https://user-images.githubusercontent.com/17170027/157748340-3cd60e25-514e-4d5e-b77a-29f018780251.png">

Author: Fixed.

Reviewer 6: I still see markup here, rather than a valid footnote link, when I view the rendered document.

Reviewer 3: Are you on the latest version? You're commenting on an outdated diff. This has been sort of fixed (the spacing issue) on the latest commit. I don't see it having a matching footnote, however.

Motivation
==========

With the change in stewardship from edX to The Center for Reimagining

> Learner, it is essential to have a well defined plan for maintaining

Reviewer 8: ```suggestion
Learning, it is essential to have a well defined plan for maintaining
```

coding platform.

The term **component** will typically mean the unit of software housed
in a single GitHub repository.  Exceptions will be made in the case of
edx-platform where there may be multiple distinguishable components in
that single repository.  Components will typically, but not always, be
hosted at the `openedx GitHub organization`_. Components will be "part
of" or "critical to" the open-source platform.

The term **maintainer** will mean a named person or persons
responsible for the health of the component and the component's community.
The remainder of this specification will be focused on the details of
the role.

The term **core contributor** will mean a named person participating
in the `Core Contributor Program`_ as a coding core contributor and
meeting the requirements of that program.

The term **maintainer scrum-of-scrums** will mean the group of all
maintainers who are responsible for flagging issues to project
leadership and proposing process improvements for component
maintenance.

The term **Technical Oversight Committee** (TOC) will mean the group
responsible for the overall oversight and stewardship of the Open edX

> project. The TOC itself was established and is governed by a formal

Reviewer 3: ```suggestion
project. The TOC's function with respect to the project is to be the
arbiter of last resort for governance matters. The TOC itself was established
and is governed by a formal
```

Reviewer 3: To address @schenedx 's concerns from a few lines up

charter.  You can read the full document `here
<https://openedx.org/wp-content/uploads/2021/12/TCRIL-Technical-Oversight-Committee-Charter-rev.11-16.21.pdf>`_.

Historical Context
------------------
Historically, the Open edX project has not had a well-defined role
of maintainer.  At times, community members stepped into this role
informally.  The role was also part of a broader "owner" role at edX.
Owners were also responsible for the health of the component they
owned both in the repository and in production; owners answered the
pager when something went wrong.

What Maintainers Do
-------------------
The role of the repository or project maintainer is both a community

> focused and a technical role [#]_.  Maintainers will be the public face

Reviewer 3: ```suggestion
focused and a technical role [#role]_.  Maintainers will be the public face
```

  component
* The ability to assess the appropriateness of proposed changes to or
  extensions of the component
* The role of the component in the overall platform and how it
  interacts with other components
* Must be available to participate in the comment period for Core Contributor
  nominees nominated to commit code to the maintainer's repo(s), as part of the
  `Core Contributor Program`_
  
It is acceptable for a component to be maintained by a team or group
as long as that arrangement does not reduce the reachability of at least one
maintainer or reduce accountability for the above responsibilites. The
maintaining team must have ways of direct contact, such as an email address and
a GitHub team, that anyone can use to get ahold of someone - it is completely
acceptable for the team to figure out arrangements for list monitoring, such as
an "on-call" rotation.

Prerequisites
-------------
In order to be a repository maintainer, you must

* Be an active member of the Open edX community who has made technical
  contributions to the project in the last 6 months.  The types of
  contributions that are relevant will match those defined in the
  `code contributors`_

> * Abide by the `Open edX Code of Conduct

Reviewer 3: ```suggestion
* Abide by the Open edX `Code of Conduct`_
```

You already have this link defined below

Becoming a Maintainer
---------------------
After the bootstrapping phase, the Open edX project will need
processes for adding maintainers, replacing maintainers, and ensuring
that new components are maintained.

Becoming a maintainer requires fulfilling the `prerequisites`_
outlined above.

A new maintainer can be added in any of the following ways:

* An existing maintainer can request that a new person become a
  component maintainer either in conjunction with the existing
  maintainer or by replacing them.  If there is no objection from
  existing maintainers, the role will be conferred.
* If the component is sufficiently complex to justify multiple
  maintainers, those maintainers form a group and should organize
  internally and present a single point of contact (like
  teams-as-maintainers above).

We are not currently limiting staff of companies from appointing,
nominating, or voting for other employees of their company.  This
would be impractical as, currently, single companies are locus of
component maintenance work. This may be addressed in future revisions
to this proposal.

Unmaintained Components
-----------------------

> In the case of unmaintained components, nominations will be made to

Reviewer 3: This is addressed below (Nominations can be made by
any member of the community including the nominee themselves.)

the scrum-of-scrums to fill the vacancy.  If there is no objection
from existing maintainers, the role will be conferred.

> If consensus cannot be reached lazily, a vote of existing maintainers

Author: Lazy consensus is a specific term of art, https://community.apache.org/committers/lazyConsensus.html.  I could be more specific about that that is intended.

> If consensus cannot be reached lazily, a vote of existing maintainers

Reviewer 3: Why not put that link in a footnote, so people can learn about the term?

> If consensus cannot be reached lazily, a vote of existing maintainers

Author: Done

  role
* A maintainer is unreachable
* A maintainer is not fulfilling the responsibilities of the role

Trigger events for removing a maintainer would be:

* A maintainer informs the project, the scrum-of-scrums, and/or the `program
  administrators`_ that they can no longer continue in the role
* A maintainer has been unreachable for more than 30 days
* A maintainer has violated the project `Code of Conduct`_
* A maintainer hasn't met project quality standards consistently
  despite feedback

In the case that a maintainer has violated the `Code of Conduct`_, the
Goverance Team, specified in the Code of Conduct, will decide what the
repercussions of the violation will be, including whether the violator
can continue in the role of maintainer.  It should be noted that
maintainers are expected to be exemplars and, as such, will be held to
a high standard.

In the case that a maintainer is not meeting project quality standards
consistently, the scrum-of-scrums will inform the maintainer with the
expectation that their components meet standards within a calendar
month.  If, after a month, the quality issues are not resolved, the

> maintainer will be removed by the scrum-of-scrums.  Additionaly, a

Reviewer 3: ```suggestion
maintainer will be removed by the scrum-of-scrums.  Additionally, a
```

> maintainer will be removed by the scrum-of-scrums.  Additionaly, a

Author: fixed.

  * How it fits into the overall platform?
  * What are its limitations and key areas of debt?
  * What new patterns or technologies should be adopted to improve the
    component?
  * How it can be extended?
* **Continuous Improvement**: Maintainers should be thinking
  about continuously improving both software and processes.

.. _program administrators:

Contact the Program Administrators
----------------------------------

Questions about the Program can be directed to the tCRIL administrator(s) via
the ``#to-be-defined`` Slack room in the `Open edX Slack
<https://openedx.slack.com/>`_ or at ``maintainer-program-admins@tcril.org``.
  

Change History
==============

01-28-2022 - `Document created <https://github.com/openedx/open-edx-proposals/pull/290/>`_

Source Material
===============

* `What Makes a Great Maintainer of Open Source Projects? <https://www.computer.org/csdl/proceedings-article/icse/2021/029600a982/1sEXoQoeO0E>`_
* `Collective Code Construction Contract <https://rfc.zeromq.org/spec/42/>`_
* `Best Practices for Maintainers <https://opensource.guide/best-practices/>`_
* `Producing OSS <https://producingoss.com/en/producingoss-letter.pdf>`_
  
Notes
=====

> .. [#] An IDA is an Independently Deployable Application, a

Reviewer 3: ```suggestion
.. [#ida] An IDA is an Independently Deployable Application, a
```

       server-side service oriented application, typically built upon
       the Django framework.

> .. [#] An MFE is a Micro-Front end, a discreet, client side

Reviewer 3: ```suggestion
.. [#mfe] An MFE is a Micro-Front end, a discreet, client side
```

       application written in Javascript, typically using the ReactJS
       framework.

> .. [#] The general acceptance of needing maintainers for OSS projects

Reviewer 3: ```suggestion
.. [#maintainer] The general acceptance of needing maintainers for OSS projects
```

       can be seen in much literature that starts from the assumption
       that maintainers are essential and instead asks, what makes a
       maintainer great, what are the best-practices maintainers
       should follow, how do we ensure maintainer don't burn out, how
       do we ensure critical parts of the software supply chain, like
       OpenSSL are maintained, etc.  I have included some revelant
       materials in the **Source Material** section.

> .. [#] Needless to say, there must be feedback loops between

Reviewer 3: ```suggestion
.. [#ops] Needless to say, there must be feedback loops between
```

       maintainers and operators of the platform and, in some cases, a
       single person will perform both roles.  Maintainers should
       certainly be concerned about the performance characteristics of
       their components and committed to their scalability.

> .. [#] In fact, studies of the traits of effective maintainers have

Reviewer 3: ```suggestion
.. [#role] In fact, studies of the traits of effective maintainers have
```

       highlighted that non-technical aspects of the role are often
       more highly valued by community members 2021 IEEE/ACM 43rd
       International Conference on Software Engineering (ICSE) `What
       Makes a Great Maintainer of Open Source Projects?

>        <https://www.computer.org/csdl/proceedings-article/icse/2021/029600a982/1sEXoQoeO0E>`_

Reviewer 3: Very much a nit but I would reorder this:

      more highly valued by community members. See `What
       Makes a Great Maintainer of Open Source Projects?
       <https://www.computer.org/csdl/proceedings-article/icse/2021/029600a982/1sEXoQoeO0E>`_
      (2021 IEEE/ACM 43rd International Conference on Software Engineering (ICSE))

===========================
OEP-55: Project Maintainers
===========================
.. list-table::
   :widths: 25 75

   * - OEP
     - :doc:`OEP-0055 <oep-0055-proc-project-maintainers>`
   * - Title
     - Project Maintainers
   * - Last Modified
     - 2022-03-16
   * - Authors
     - Edward Zarecor <e0d@tcril.org>, Sarina Canelake <sarina@tcril.org>
   * - Arbiter
     - Jeremy Bowman <jbowman@2u.com>
   * - Status
     - Draft
   * - Type
     - Process
   * - Created
     - 2022-01-28
   * - Review Period
     - 2022-03-21 - 2022-04-11
   * - Resolution
     - <links to any discussions where the final status was decided>
   * - References
     - <links to any other relevant discussions or relevant related materials>

.. contents::
   :local:
   :depth: 3

Abstract
========

The Open edX platform is a service oriented platform for authoring and
delivering online learning based on the `Django framework`_.  At the
highest level, the platform is composed of numerous services (IDAs

> [#ida]_), micro-front ends (MFEs [#mfe]_) and libraries.  The code for these

Reviewer 9: nit: "micro-front ends" ➡️ "micro-frontends"

services and libraries is managed in git and hosted on the GitHub
platform.

Each component part of the Open edX platform requires maintenance to
ensure that it meets, and continues to meet, the project definition of
high quality.  That definition will include project standards,

> appropriate documentation,and fitness for use, among other things.

Reviewer 10: ```suggestion
appropriate documentation, and fitness for use, among other things.
```

That definition will evolve as project needs do and we learn what
works and what doesn't.

Component maintenance is work.  To ensure that that work is done, a
person or persons must be recognized as responsible for doing it,
and held accountable for its performance.

The Open edX project, like many projects, calls this role
"maintainer."  The notion of a project maintainer is typical of
open-source projects and widely considered a best practice [#maintainer]_.

The maintainer role is a leadership role that encompasses both
technical and community management elements.  Both dimensions of the role are
equally important.

All non-deprecated components of the Open edX project must have one
or more named maintainers.  When no such maintainer(s) exist, this will
be considered an urgent problem to solve.

This OEP proposes a formal definition of the role of maintainer that
will be open to all qualified Open edX community members.  Maintainers
will be responsible for the health of platform components, but not for
their performance in any particular community member's production

> environment [#]_.

Reviewer 3: I don't see a matching footnote for this footnote. Further, could we please continue using named footnotes? 

```suggestion
environment [#prodenv]_.
```

way of ensuring this is to have a named person or persons who are
responsible for getting the work done, a measurable definition of
"well-maintained," and feedback looks to ensure accountability and
continuous improvement of our processes.

Specification
=============

Definitions
-----------

The term **repository** will mean a repository on the GitHub social
coding platform.

The term **component** will typically mean the unit of software housed
in a single GitHub repository.  Exceptions will be made in the case of
edx-platform where there may be multiple distinguishable components in
that single repository.  Components will typically, but not always, be
hosted at the `openedx GitHub organization`_. Components will be "part
of" or "critical to" the open-source platform.

The term **maintainer** will mean a named person or persons
responsible for the health of the component and the component's community.
The remainder of this specification will be focused on the details of
the role.

The term **core contributor** will mean a named person participating
in the `Core Contributor Program`_ as a coding core contributor and

> meeting the requirements of that program.

Author: I'm open to suggestions about how to clarify.  I see maintainers as the ultimate responsible party.  I think that some things we ask CCs to do today is really maintainer work, but we didn't have that notion defined.

> meeting the requirements of that program.

Reviewer 3: I feel like the maintainer is the organizer and backstop. If a CC could commit to say Repo A and Repo B, I wouldn't necessarily expect them to be constantly on top of requirements updates and needs (although definitely somewhat!). But I'd expect a maintainer to have the "big picture" of Repo A's overall health in mind, and do a combo of doing the upgrade work as well as farming it out to Repo A's CCs. @antoviaque do you agree or have something else to propose? What I wrote is just my current thoughts and am open to reshaping them.

> meeting the requirements of that program.

Reviewer 11: @sarina Yup, what you mention makes sense to me :+1:

> meeting the requirements of that program.

Author: I think we're all in agreement on this and will resolve.

The term **maintainer scrum-of-scrums** will mean the group of all
maintainers who are responsible for flagging issues to project
leadership and proposing process improvements for component
maintenance.

The term **Technical Oversight Committee** (TOC) will mean the group
responsible for the overall oversight and stewardship of the Open edX
project. The TOC's function with respect to the project is to be the
arbiter of last resort for governance matters. The TOC itself was established
and is governed by a formal
charter.  You can read the full document `here
<https://openedx.org/wp-content/uploads/2021/12/TCRIL-Technical-Oversight-Committee-Charter-rev.11-16.21.pdf>`_.

Historical Context
------------------
Historically, the Open edX project has not had a well-defined role
of maintainer.  At times, community members stepped into this role
informally.  The role was also part of a broader "owner" role at edX.
Owners were also responsible for the health of the component they
owned both in the repository and in production; owners answered the
pager when something went wrong.

What Maintainers Do
-------------------
The role of the repository or project maintainer is both a community

> focused and a technical role [#role]_.  Maintainers will be the public face

Reviewer 3: To me this feels like the role that the Core Contributor - Project Manager is meant to address. I know Jorge is working with a goal of eventually becoming a CC in this role. And again to my earlier point - I don't love the idea of a club that has a high enough bar that people in certain roles cannot hope to achieve the required monthly status - that makes some roles feel lesser than others. A tiered system also has that feeling. I would like to discuss how we can be inclusive and friendly, while asking for and celebrating higher commitments of some people, groups, or ecosystem partners.

> focused and a technical role [#role]_.  Maintainers will be the public face

Reviewer 3: Hi @antoviaque @adamstankiewicz - is this thread a blocking concern for merging this OEP to provisional status?

Personally I am ok with the current wording. "Technical" does not need to mean you write code. It does mean you understand the technical ins & outs of GitHub, for example. It may mean you understand how to debug RST or work in other domain-specific areas. What do you think? I would like either a "this is ok as-is and we can iterate as needed", or a proposed wording change. Thanks!

> focused and a technical role [#role]_.  Maintainers will be the public face

Reviewer 11: @sarina Since it's already a large OEP with a lot of changes, I'm fine with merging as-is for now. After all, that's also how we started it for core committers, by focusing on the technical/dev roles. Once we have seen a bit how maintainers work out on a smaller set of duties, we can always extend it then to other roles if it looks pertinent.

of each component and should be welcoming and available.  While
maintainers should strive to always be friendly and kind in their
communication, firmness and directness are equally important.
Maintainers must enforce the project's standards.  In cases where a
proposed change isn't aligned with the our standards or direction, it
is best to make that clear as early as possible.

Non-technical responsibilities of maintainers include:

* Effective communication with community members, other maintainers,
  and project leadership
* Selection and approval of Core Contributors with commit rights to their repositories
* Disciplined management to ensure that maintenance work is done
* Enforcement of the project's standards, both technical and social: maintainers
  must remain friendly, kind, and open, but that should not supercede

>   enforcement of technical standards

Reviewer 3: @antoviaque @kdmccormick how does this wording sound?

> Enforcement of the project's standards, both technical and social: maintainers and project discussions must remain friendly, kind, and open, but that should not supersede enforcement of technical standards

>   enforcement of technical standards

Reviewer 3: I like your rewrite, I'll use that.

I would feel more comfortable with a clause like, "Ensuring that project discourse occurs in public spaces, enabling Open edX contributors to read and comment on discussions. Use discretion in cases that may need to remain private, such as security disclosures." 

Qualifying the type of discourse we think may need to be private is better than giving a blank check. I hear what you're saying about early ideation, but that is a blurry line I'd rather not call out.

* The ability to solve or escalate issues facing users of the

>   component

Reviewer 12: This sounds like a technical responsibility? And, instead of or in addition to this, I would suggest: "Ensure that bug reports are read, acknowledged, and prioritized in a timely fashion"

Technical responsibilities of maintainers include:

* Understanding the current component architecture and implementation

> * Ensuring that basic documenation for the component exists and is

Reviewer 12: Typo: `documenation`

  accurate
* Having a documented technical vision for the evolution of the
  component
* The ability to assess the appropriateness of proposed changes to or

>   extensions of the component

Reviewer 11: NB: This is related to one of the items of feedback from a core contributors sprint retrospective - see section _"2. Getting feedback/reviews for OSPRs is as always a challenge"_ from https://discuss.openedx.org/t/core-contributor-sprint-retro-march-4-18th/6876/4

>   extensions of the component

Reviewer 4: > commit to provide prompt reviews for changes

On one hand, this would be an amazing state to end up in. On the other hand, from my experience on a team that "maintained" a couple dozen repos, reviewing a steady stream of PRs from outside my team felt like an enormous and overwhelming task.

Personally, I would be happy if maintainers promptly _triaged_ all incoming PRs (which might include giving a time estimate and/or advertising the PR to other willing reviewers) and committed to reviewing a finite number of PRs per maintained repo per period of time (strawman example: "maintainers should triage all incoming PRs, and themselves review an average of >=1 PR per component per month"). Committing maintainers to entirely emptying their review backlog, though, might be too big of an ask at this point in time.

>   extensions of the component

Reviewer 4: > commit to provide prompt reviews for changes

On one hand, this would be an amazing state to end up in. On the other hand, from my experience on a team that "maintained" a couple dozen repos, reviewing a steady stream of PRs from outside my team felt like an enormous and overwhelming task.

Personally, I would be happy if maintainers promptly _triaged_ all incoming PRs (which might include giving a time estimate and/or advertising the PR to other willing reviewers) and committed to reviewing a finite number of PRs per maintained repo per period of time (strawman example: "maintainers should triage all incoming PRs, and themselves review an average of >=1 PR per component per month"). Committing maintainers to entirely emptying their review backlog, though, might be too big of an ask at this point in time.

>   extensions of the component

Reviewer 3: @antoviaque @kdmccormick I agree with both of you. I have personally always been a fan of closing PRs if review was not going to happen to them, which to me does fall under "review promptly" (as a person on the other end of a PR, it's so frustrating to _not know_ and/or get pinged six months after you've dropped context... I'd rather just get a No). But Kyle's suggestion of "triage" is much nicer. Can we agree on this wording?

> The ability to promptly triage incoming requests that propose changes to or extensions of the component, assessing their appropriateness and/or routing them to proper reviewers

>   extensions of the component

Reviewer 4: Works for me 👍🏻

>   extensions of the component

Reviewer 11: @sarina Sounds good to me as a first step -- this way we'll get to see what happens, and having a reliable and fast triage would be a great first step. If there are still delays with _actually_ getting contributions reviewed afterwards though, I would require actual reviews from maintainers within a given timeframe (2 weeks to start?). As an open source project, being able to promptly review new contributions is important - we won't be a mature project until we can offer it reasonably reliably and promptly.

* The role of the component in the overall platform and how it

>   interacts with other components

Reviewer 12: This bullet point doesn't seem to make sense in the context of this list of responsibilities ("Technical responsibilities...include...the role of the component"). I think a verb is missing?

* Must be available to participate in the comment period for Core Contributor
  nominees nominated to commit code to the maintainer's repo(s), as part of the
  `Core Contributor Program`_

>   

Reviewer 12: The section on technical responsibilities doesn't explicitly mention a few things that come to mind when I think of "maintainer" responsibilities:
* Something like "ensure ongoing compatibility with other components of the Open edX ecosystem as they are developed" (e.g. changes to edx-platform, or Paragon, or even new Ubuntu LTS versions, etc.). This could be part of "Having a documented technical vision for the evolution of the component"?
* Something like "ensure dependencies are updated as needed to avoid security risks or use of components that unsupported/unmaintained/end-of-life."

I suppose those could fall under "Disciplined management to ensure that maintenance work is done" and "enforcement of technical standards" from the non-technical section? Either way if they're expected I think calling them out explicitly is helpful.

Reviewer 3: @bradenmacdonald this has now been added:

```
* Ensure that dependencies are updated as needed to avoid security risks or use
  of components that or not fit for use, e.g., unsupported, not maintained, or
  at their end-of-life.
```

To this point:

> Something like "ensure ongoing compatibility with other components of the Open edX ecosystem as they are developed" (e.g. changes to edx-platform, or Paragon, or even new Ubuntu LTS versions, etc.). This could be part of "Having a documented technical vision for the evolution of the component"?

I'm also not sure where to put this or how to phrase this. I could see simply putting your quote under Technical Responsibilities. Do you have a suggestion/opinion?

Reviewer 12: Putting it under technical responsibilities seems fine to me!

It is acceptable for a component to be maintained by a team or group
as long as that arrangement does not reduce the reachability of at least one
maintainer or reduce accountability for the above responsibilites. The
maintaining team must have ways of direct contact, such as an email address and
a GitHub team, that anyone can use to get ahold of someone - it is completely
acceptable for the team to figure out arrangements for list monitoring, such as
an "on-call" rotation.

Prerequisites
-------------
In order to be a repository maintainer, you must

* Be an active member of the Open edX community who has made technical
  contributions to the project in the last 6 months.  The types of
  contributions that are relevant will match those defined in the
  `code contributors`_
* Abide by the Open edX `Code of Conduct`_
  <https://openedx.org/code-of-conduct/>`_
* Be a member of the `Core Contributor Program`_ with rights to the repository

>   in question who has demonstrated excellence in the role of core contributor.

Reviewer 3: First off, legally, to get write access to a repo, one will need to formally be in the Core Contributor program.

Secondly, I've been reading about communities. I'm not very comfortable having a program that feels so exclusive, or a program that has such strict standards that people lose status easily. This is a topic I would like to discuss further and I think in person would be good. See https://discuss.openedx.org/t/conf-bof-governance-topics/6841 and in there a reference I quote.

>   in question who has demonstrated excellence in the role of core contributor.

Reviewer 3: OK, I definitely hear you and agree with many of your sentiments - I look forward to discussing more at the conference.

My concern (as I also raised in a discourse thread about granting write access to WG repos) is proliferating different legal agreements and expecting someone (particularly, myself and the legal team at tCRIL) to monitor everyone's activities and what they're doing with their permissions, and have them sign new agreements when they move into different categories, possibly having to invalidate old agreements... it becomes a headache to administer. So - perhaps selfish, but also (given the burden on our legal counsel already) realistic.

>   in question who has demonstrated excellence in the role of core contributor.

Author: @antoviaque do you think landing this conversation is blocking for merging this?  I do agree when need to find and strike the right balance between opportunities to participate and accountability to contribute.  This feels like a great conversation to have at the conference as part of the Friday session on shared ownership.

>   in question who has demonstrated excellence in the role of core contributor.

Reviewer 11: @e0d I think it would be useful to clarify it yes -- the core contributor program is still quite young, so it will be important to have clear definitions of both programs and status, so we don't end up with confusion about it. Imho it needs to be clear what the incentive for organizations to keep their core contributor commitment is, vs maintainership -- especially if one of the two roles have a lower threshold.

Discussing it at the conference sounds good! I'll suggest the topic in the thread @sarina started for collecting governance topics to discuss (https://discuss.openedx.org/t/conf-bof-governance-topics/6841)

>   in question who has demonstrated excellence in the role of core contributor.

Author: @antoviaque and @sarina I'm going to attempt to summarize our conversation.

We agreed that we're going to keep the requirement that one must be a Core Contributor to become a Maintainer for now.  If we find that that limits participation in practice -- not a goal -- we'll revisit.

We value the level of commitment that comes with being a Core Contributor and don't want to reduce that.  We want to ensure that the programs are congruent and not working at cross purposes.  Hours spent maintaining components would obviously count toward a Core Contributors commitment.  

If we see issues that need to be addressed we can propose amendments to this OEP as we learn.

>   in question who has demonstrated excellence in the role of core contributor.

Reviewer 3: @e0d - yes, that is what I took away from our conversation. Let's see how this works, and adjust if necessary.

  From time-to-time, in the case of a strong candidate and project need, core
  contributor status and maintainer status may be conferred at the same time.
* In the case of repos being moved into the `openedx GitHub organization`_, be
  an original maintainer of the code.

Bootstrapping
-------------

> In order to ensure an orderly roll-out of the formal maintainer role,

Reviewer 10: I'm writing this comment just to explicitly say out loud what I believe our shared expectations already are.  I imagine they may not be obvious to folks who haven't been intimately involved in the edX -> tCRIL transition process.

The sentence here about roll-out for existing 2U staff makes sense to me, under the assumption that - as @sarina says below - we expect to codify the exact timelines and compliance expectations elsewhere in a more nuanced way.  I don't think it's anyone's expectation that the moment we switch from "owners" to "maintainers" we're going to start a clock and start removing maintainers one month later for not bringing their repositories up to compliance with all our standards.  

Rather, the transition has always had an expectation that we will "do no harm" to the edX business as we transition edX from the "owner" to a true peer maintainer.  In fact, I'd expect this goes for other businesses too as they onboard to maintainership.

I would hope that transitioning 2U staff from owners to maintainers is unconcerning and almost a no-op in the short term for those of us on the 2U side, but that we apply pressure in the medium-term to ensure we prioritize bringing the platform - regardless of maintainer - up to our standards.  Once we have a reasonable expectation of quality in most quarters, then I'd expect the "one month" clock to start coming into play.

Again, my assumption is that none of that is controversial, but isn't called out explicitly here and probably doesn't need to be.

Author: Thanks @davidjoy I agree with your framing.  We have flexibility in terms of when we expect things to be done and the idea is not to create an overwhelming body of new work.  The transition needs to be orderly and I do believe that much of the needed work is being done today.  Will there be some new work?  Well yes, there will be a little, but I think it will benefit all users of the project and set the stage of distributed, shared maintenance.

Reviewer 3: Agree @davidjoy . I want to double-check that there's no action needed on this comment?

Reviewer 10: 👍🏻 Yeah!  No action needed.

The Maintainer Scrum-of-Scrums
------------------------------
In order to coordinate certain maintainer responsibilities including
voting on nominated maintainers, iterating on standards, and removing
maintainers, minimal mechanisms for coordination are required.

For now, we will assume that maintainers will self-organize to
complete the critical activities.

Coordination will typically be over asynchronous channels.

From time-to-time, a task force may be required to complete certain
project-wide responsibilities like developing repository standards,
as well as tools for measuring compliance to those standards.

Becoming a Maintainer
---------------------
After the bootstrapping phase, the Open edX project will need
processes for adding maintainers, replacing maintainers, and ensuring
that new components are maintained.

Becoming a maintainer requires fulfilling the `prerequisites`_
outlined above.

A new maintainer can be added in any of the following ways:

* An existing maintainer can request that a new person become a
  component maintainer either in conjunction with the existing
  maintainer or by replacing them.  If there is no objection from

>   existing maintainers, the role will be conferred.

Reviewer 3: I agree. However I think it may make sense that only coding core contributors have voting/veto rights. Thoughts @antoviaque ?

>   existing maintainers, the role will be conferred.

Reviewer 11: @sarina Technical expertise is important to judge the technical capability of a maintainer -- however technical ability is not the only criteria a candidate maintainer has to meet, so even if a core contributor might be unable to judge the technical merits of the candidate, they might have legitimate non-technical objections, about things like social interactions, trust-worthiness, etc.?

>   existing maintainers, the role will be conferred.

Reviewer 3: Ah. I was thinking it would be much like the CC process. My understanding of that process is that only CCs have affirmitive voting rights, but anyone bringing up a concern will be listened to and the nomination process may be halted because of those concerns.

So I misspoke! Would it be reasonable to adopt that model? CCs are the ones who count as "voters" for the +1s, but the community at large is welcome to participate - and will be heard? From OEP-54:

> All community members are welcome to provide comments on the thread.

> CoC violations or other behaviors not upholding the “Conduct” quality of the 3 Cs ought to be raised by anyone, whether in the Discourse thread or to the Program Admins.

> If anyone has concerns they feel cannot be raised publicly, they should [Contact the Program Administrators](https://open-edx-proposals.readthedocs.io/en/latest/processes/oep-0054-core-contributors.html#contact-the-program-administrators) to determine how to proceed. The Administrators should discuss the concerns and try to figure out a way the person could post those concerns themselves; if they cannot, the Administrators should post an anonymized version of the concerns on the Discourse thread.

>   existing maintainers, the role will be conferred.

Reviewer 3: @e0d - I think I would change this a bit. I think _anyone_ should be able to object. If you are saying there's no affirmative process (such as the core contributor 5 votes), I would change this to:

> If there is no objection from existing maintainers, core contributors, or the community, the role will be conferred.

I think it's important that someone in the community be able to bring up a concern - they shouldn't be discounted just because they're not in more of a leadership position.

>   existing maintainers, the role will be conferred.

Reviewer 4: @sarina The [New CC confirmation process](https://github.com/openedx/open-edx-proposals/blob/master/oeps/processes/oep-0054-core-contributors.rst#new-core-contributor-nomination-process) allows anyone to comment, but reserves the right to veto for only existing Core Contributors. I like this model, because it allows anyone to express concern, but protects us from some troll/griefer coming in and voting No on everything just to mess with our process. In the cases where a non-CC's concern is valid, then I'm confident that a CC would support it by issuing a No vote on their behalf.

Tangentially, I also like that the CC nomination process if very specific in where it should be done and how long it should take.

Can we adopt a similar stance for Maintainership? For example:
> * An existing maintainer or core contributor can request that a new person become a component maintainer either in conjunction with the existing  maintainer or by replacing them. The request should be publicized in Discourse, available for all community members to comment their support, questions, and/or concerns. In particular, existing maintainers and core contributors may object to the proposed change. After one week, if there are no objections from existing maintainers or core contributors, the role will be conferred.

>   existing maintainers, the role will be conferred.

Reviewer 3: I don't mind that. I'd like @e0d to give the final signoff on this point.

>   existing maintainers, the role will be conferred.

Reviewer 3: Actually... I've just pushed a change to address this which Ed can review. I don't think this is a controversial change - it lines up maintainership nomination with CC nomination process, and one process is better than two.

https://github.com/openedx/open-edx-proposals/pull/290/commits/26b21cc8f2600192acde8906ecaf1c5367c53b18

* If the component is sufficiently complex to justify multiple
  maintainers, those maintainers form a group and should organize
  internally and present a single point of contact (like
  teams-as-maintainers above).

We are not currently limiting staff of companies from appointing,
nominating, or voting for other employees of their company.  This
would be impractical as, currently, single companies are locus of
component maintenance work. This may be addressed in future revisions
to this proposal.

Unmaintained Components
-----------------------
In the case of unmaintained components, nominations will be made to
the scrum-of-scrums to fill the vacancy.  If there is no objection
from existing maintainers, the role will be conferred.

If consensus cannot be reached lazily, a vote of existing maintainers

> will be taken.  Maintainers can approve a nomination with the approval

Reviewer 11: @sarina @e0d To also apply my comment from [this related thread](https://github.com/openedx/open-edx-proposals/pull/290#discussion_r869429511) here too, could we also apply this diff at the same time?

the scope of this document.  For example, here we will note that
maintainers are responsible for ensuring that repositories meet
project standards, but not the details of those standards.  Those
standards - including establishing a channel for feedback, and details on how
and when that channel will be monitored - will be codified in an OEP over time,
but will initially be defined in the wiki page `requirements for public
repositories`_ to allow rapid iteration.

The key aspects of the maintainer role fall into the following
categories

* **Community Stewardship**: Maintainers are representatives of the
  project in the community and are responsible for contributing to an
  environment that is welcoming, vibrant, and growing.
* **Project Management**: Maintainers are responsible for ensuring
  that critical maintenance is done, but this doesn't mean doing it all
  themselves. Effective delegation is an important maintainer skill.
* **Quality Assurance**: Maintainers are responsible for the quality
  of their components.  Quality will mean meeting defined project
  standards.  It can also mean improving our standards or improving
  our processes.

> * **Technical Vision**: Maintainers should have a documented technical

Author: I think that that should be captured as well.  When we think about a README for any component, we'd typically want to know why the thing exists and how it is engineered.  I believe that both can be captured very succinctly. The technical vision might be a little more specific on details.

> * **Technical Vision**: Maintainers should have a documented technical

Reviewer 3: @adamstankiewicz this is another place where I think the word "technical" may be a hang-up for your team, whereas I don't think your team should consider yourself un-technical. Are you able to provide a vision (technical or otherwise) that answers the questions that follow?

```
  * What is the component for?
  * How it fits into the overall platform?
  * What are its limitations and key areas of debt?
  * What new patterns or technologies should be adopted to improve the
    component?
  * How it can be extended
```

Is there a rewording or something missing here that you'd like to see?

> * **Technical Vision**: Maintainers should have a documented technical

Reviewer 9: We definitely wouldn't consider the Paragon team/component un-technical, but wanted to surface that Paragon goes beyond a technical vision alone. For example, the design vision for Paragon and how we do design across the organization helps inform/lead some aspects of our technical vision for Paragon (and vice versa), likely why I was getting hung up on explicitly calling it a technical vision alone (i.e., the vision for components like Paragon are all-encompassing of both a design and technical vision that coincide together).

The technical vision for Paragon, as a result, then would actually be more of a "here's where we're going with design and, as such, here's what we need to do technically", for example.

If I were to suggest a change in the wording, it'd likely be referring to a "Component Vision" versus a "Technical Vision" specifically. But, don't consider this as blocking feedback; I don't feel strongly about the word choice as either way, the technical vision for Paragon will include aspects of the design vision in order to account for the aforementioned questions.

> * **Technical Vision**: Maintainers should have a documented technical

Reviewer 3: @adamstankiewicz I actually think that's a good point - I would absolutely want a component's vision to be comprised not just of future technical architecture work but where the component is going (ie, what's on its product roadmap). In the interest of getting this resolved quickly I'll touch base with Ed on this point - would you think a "technical + product vision" wording encompass your thoughts?

> * **Technical Vision**: Maintainers should have a documented technical

Reviewer 3: Ok @adamstankiewicz Ed & I have come to an agreement on our thoughts here, being that, eventually we would love to see Maintainers document both a technical and a design/product vision. The current state of most components is that latter vision is not well formed/well available, and we're currently working on some of that via the Product Roadmap and Product Working Group.

Ultimately, we're going to choose to leave the wording as-is, but we would LOVE if components (especially ones like Paragon) started moving towards the direction of capturing the entire component vision. And after we roll this out, I think we'll be ready to do an iteration on this OEP that does bring the product/design vision more closely into the fold.

> * **Technical Vision**: Maintainers should have a documented technical

Reviewer 13: This makes sense to me, the goal here is not to define everything a maintainer is *allowed* to do but rather the things they *must* do.  For components like paragon, that are ahead of the curve, I think doing more and showing the value of having product vision is a great way to lead by example to push our standards even further in the future.

===========================
OEP-55: Project Maintainers
===========================
.. list-table::
   :widths: 25 75

   * - OEP
     - :doc:`OEP-0055 <oep-0055-proc-project-maintainers>`
   * - Title
     - Project Maintainers
   * - Last Modified
     - 2022-03-16
   * - Authors
     - Edward Zarecor <e0d@tcril.org>, Sarina Canelake
       <sarina@tcril.org>, Xavier Antoviaque <antoviaque>, Olivia
       Ruiz-Knott <oliviaruizknott>, Ned Batchelder <nedbat>, David
       Joy <davidjoy>
   * - Arbiter
     - Jeremy Bowman <jbowman@2u.com>
   * - Status

>      - Draft

Reviewer 4: Should be Provisional or Accepted, unless you plan to merge a Draft of this.

Author: I think that the status update would happen just prior to merging.  I would defer to @jmbowman 's recommendation as to status.

Reviewer 14: My $0.02 would be Provisional but with a concrete timeline / set of objectives for what constitutes "implemented". That might be:

- Set of repos included in Open edX is known and clearly enumerated via some single source of truth
- Maintainers are declared for each of the components of Open edX (and have responded accepting the responsibilities)
- A system is in place for providing maintainers with quant / qual feedback on their progress "maintaining project standards"
- Scrum of scrums is meeting

FYI @e0d 2U is trying to think about pilot repos as part of [our rollout plan](https://github.com/edx/open-source-process-wg/issues/31) as well, so it would be good to coordinate a bit and make sure we have a variety of cases across different orgs.

Author: This feels like the right list of priorities in the pilot phase.

Reviewer 3: I made this wiki section to help us document pilot phases & timelines. Could we ensure that documentation is put here? https://openedx.atlassian.net/wiki/spaces/COMM/pages/3426418790/Maintainership+of+Open+edX+Repos

==========

With the change in stewardship from edX to The Center for Reimagining
Learning, it is essential to have a well defined plan for maintaining
the platform.

In order to ensure that the project is sustainable for the long term,
maintenance must be shared by the entire community.  The project must
be robust enough that it can tolerate firms and contributors leaving
the ecosystem.

As a project we must acknowledge that maintenance is hard work, but
work that must be done for the platform to be fit for use.  The best
way of ensuring this is to have a named person or persons who are
responsible for getting the work done, a measurable definition of
"well-maintained," and feedback looks to ensure accountability and
continuous improvement of our processes.

Specification
=============

Definitions
-----------

The term **repository** will mean a repository on the GitHub social
coding platform.

The term **component** will typically mean the unit of software housed
in a single GitHub repository.  Exceptions will be made in the case of
edx-platform where there may be multiple distinguishable components in
that single repository.  Components will typically, but not always, be

> hosted at the `openedx GitHub organization`_. Components will be "part

Reviewer 4: What repositories are you thinking we'd want maintainers for that are outside of the `openedx` organization?

Based on the wording of **Bootstrapping**, I think this OEP would require maintainers for all public repos in the edx organization. Do we want to go down that path?

The community also depends on repos in the `overhangio` organization, namely tutor and its official plugins. I believe @regisb has established his own system of maintainers, though, so I assume  that those repos don't fall in the scope of this OEP.

Author: I want to provide for things like hooks and signals, Blockstore, etc.  I realized those have migrated into the openedx org, but there are recent examples of critical components that were not in that org.  I want to make sure that if and when that case arises we have congruent expectations around maintenance.

Author: Resolving for future discussion in OEP-57.

of" or "critical to" the open-source platform.

The term **maintainer** will mean a named person or persons
responsible for the health of the component and the component's community.
The remainder of this specification will be focused on the details of
the role.

The term **core contributor** will mean a named person participating
in the `Core Contributor Program`_ as a coding core contributor and
meeting the requirements of that program.

The term **maintainer scrum-of-scrums** will mean the group of all
maintainers who are responsible for flagging issues to project
leadership and proposing process improvements for component

> maintenance.

Author: I don't see this change as valuable.  Can you help me understand why virtual is important?

> maintenance.

Author: Got it, I've made the proposed change.

Historical Context
------------------
Historically, the Open edX project has not had a well-defined role
of maintainer.  At times, community members stepped into this role
informally.  The role was also part of a broader "owner" role at edX.
Owners were also responsible for the health of the component they
owned both in the repository and in production; owners answered the
pager when something went wrong.

What Maintainers Do
-------------------
The role of the repository or project maintainer is both a community
focused and a technical role [#role]_.  Maintainers will be the public face
of each component and should be welcoming and available.  While
maintainers should strive to always be friendly and kind in their
communication, firmness and directness are equally important.
Maintainers must enforce the project's standards.  In cases where a
proposed change isn't aligned with the our standards or direction, it
is best to make that clear as early as possible.

Non-technical responsibilities of maintainers include:

* Effective communication with community members, other maintainers,
  and project leadership

> * Selection and approval of Core Contributors with commit rights to

Reviewer 15: [I've expressed these thoughts to others and was encouraged to post here!]

I'm a recent addition to edX/2U, and upon joining last year I was pleasantly surprised (and, honestly, terrified in a positive way) that I was able to merge commits to `master` in any repo.

I discovered that this was very empowering in several ways:
* I could offer manpower to perform deployments for other teams 
* I could revert other team's mistakes easily when on call
* I could unblock others by approving PRs for repos my team doesn't explicitly own but in which I happen to have an expertise in
* I could substitute for others when they were were unavailable
* (and vice-versa: people very often offer/revert/unblock/sub for me, and I sense this is a culture in the Open edX community in general!)

These capabilities are also vital given:
* How short-staffed edX/2U (and I'm sure everyone else) is. For example, two teams I've happened to have occasion to do the above for had and still have zero developers on team.
* How interconnected the Open edX codebase is. I have found bugs in my team's repo are caused by commits made 1-3 non-owned repos away. 

I am confused about the ramifications of this OEP to our ability to offer/revert/unblock/sub for others. This is because:

* I have heard mixed interpretations from others as to whether and to what extent this OEP will impact: (a) commit rights, (b) merge rights, and (c) PR approval requirements.
  * For example, in the most extreme case: Will the ability to create & commit to a new branch in a repo be on a repo-by-repo basis, as determined by the maintainer?
* I do not understand how this OEP and the Core Contributor OEP accommodate the fact that many maintainers are organizations or teams within organizations rather than individuals.
  * For example: Would a new employee of an organization be eligible to be a maintainer or even a Core Contributor? If not, would they be unable to fully ship code until they are eligible to become a Core Contributor?
* It seems like, at least initially, some repositories may need a very large number of maintainers. `edx-platform`, e.g.

Would someone be willing to help me gain clarity on this?

-------

A separate thought experiment: Why not add all Core Contributors to `pull-push-all` instead of restricting commit/approve/merge/deploy repo-by-repo?

Put another way:
* I don't understand why maintainers need a hard ACL control over which Core Contributors have commit privileges to their repo.
* I'd like to suggest allowing all Core Contributors to commit to all repos increases rather than decreases developer agency, for edX/2U but especially for the broader community.
   * (I'd appreciate input here especially if this is not right, as I'm not a member of the broader community.)

-------

I am sure I may be making a number of incorrect assumptions or am lacking some understanding due to error/skimming/newness/missing info/etc. I'm looking forward to hearing what they may be!

Reviewer 4: @pshiu 's - I really appreciate your take here. You've made me step back and reflect on what the goals of our commit access control system should be. I can personally only think of four requirements:
1. Ensure that each contribution is reviewed by at least one person with the relevant context and then merged by a core contributor (who may or may not be the author).
2. Cause little-to-no undue friction for development and on-call responsibilities.
3. Provide a consistent & level playing field for the whole community.
4. Be easy to understand and administrate.

(Note: I didn't include anything about "quality" because it's so subjective and IMO will stem naturally from goals 1-3 in conjuction with the existing criteria for becoming a core contributor in the first place)

I do think the system we have been using for non-2U community members via the CC program could meet goals 1-4 if rolled out to 2U. That being said, the system you describe also meets those goals, are could be even stronger on goals 2 and 4!

As for goal 1 (ensuring review), perhaps that is where repo maintainers could make the judgement call on what "relevant context" is, using a tool like codeowners, main branch protection, or simple documentation to enforce review requirements.

I am curious what others think!

Author: Apologies in advance for this long response.  I would like to resolve this comment this week and be in a position to merge a **provisional** version of this OEP on Friday :slightly_smiling_face: 

Phil thanks for your response.  It’s clear that a lot of time and thought went into it.

I’d like to respond to a few of your points.  

Firstly, I don’t think that being able to “fully ship code” and “ship code instantly with only myself to decide” are the same thing.  Developer velocity is certainly a priority, but it’s not the only one.

Giving all core contributors universal, unfettered access is an interesting proposal.  However, I don’t think it’s the right choice.  It is standard practice in open source projects - and many closed source ones as well -  that rights to commit to or maintain the software are granted based on evidence.  Over time contributors demonstrate that they have expertise, are reliable, and have good judgment.

One of the critical jobs of the proposed maintainer role is to ensure that the software is of high quality.  I believe that we should be, for example, minimizing rollbacks rather than optimizing for them.  I do think we’ll need to have an affordance, say 2U admins with special privileges, to ensure that 2U operations remain stable and fast.  But 2U are not the only ones running the software in production.  So we’ll have to consider whether other organizations should also have such admins.  

Ultimately, I would like to see 2U have the same privileges as the rest of the community.  I think that it makes sense for this to happen by decreasing 2U’s default privileges, not by elevating everyone else’s to a point that I don’t think is safe or supported by best practices. We are in parallel working on a pilot of decreasing permissions (“scoped access permissions pilot”), separate from this OEP. This OEP does not impact the areas you mention: (a) commit rights, (b) merge rights, and (c) PR approval requirements. This OEP is about how to make sure all repos are maintained/have high quality standards.

2U – as other community members always have needed to do – will have to have some operational procedures in place that ensure safety, stability, and velocity.  I know that Ned among others has been thinking about this already. To your questions, “Would a new employee of an organization be eligible to be a maintainer or even a Core Contributor? If not, would they be unable to fully ship code until they are eligible to become a Core Contributor?” - this will be tackled with the scoped access permissions pilot. Let’s hold this conversation for then.

Regarding the complexity that is edx-platform, the OEP provides an affordance for teams as maintainers and also indicates that for repositories like edx-platform, parts of it can be maintained by distinct people or teams.

Reviewer 15: Thanks for the response, @e0d! Like you said, I'm sure these concerns will be tackled in the scoped access permissions pilot, so looking forward to following the discussions there.

Goverance Team, specified in the Code of Conduct, will decide what the
repercussions of the violation will be, including whether the violator
can continue in the role of maintainer.  It should be noted that
maintainers are expected to be exemplars and, as such, will be held to
a high standard.

In the case that a maintainer is not meeting project quality standards
consistently, the scrum-of-scrums will inform the maintainer with the
expectation that their components meet standards within a calendar
month.  If, after a month, the quality issues are not resolved, the
maintainer will be removed by the scrum-of-scrums.  Additionally, a
maintainer who is warned three (3) times in a single calendar year
will also be removed. The `program administrators`_ serve as a
backstop - any community member may contact them at any time if
a problem arises and the normal channels of communication are
not working.

Documenting Maintainers
-----------------------
As it is essential that maintainers are discoverable and reachable,
the person or persons responsible for maintaining a component should
be listed in the repository README file along with the mechanism for

> contacting a maintainer.

Reviewer 4: At edX there developed an idea of "owning ownership". That is, someone needed to define what ownership meant, keep the documentation of owners up-to-date, and maintain repository health monitoring mechanisms, and facilitate ownership transfers. The grand Ownership Spreadsheet didn't happen until a few folks (such as @feanil and the rest of Arch-BOM) took on that mantle of meta-ownership.

I think this OEP needs a similar person or group of people to, uh, "maintain maintainership". Will that be you, @e0d? Or perhaps tCRIL, or a particular working group?

Author: I think this is a service that tCRIL provides or ensures is provided.  Ideally the mechanisms are owned by the whole community, but it's clear that it's someones job at the end of the day.
